src/dodal/devices/tetramm.py

```diff
@@ -0,0 +1,278 @@
+
+    @property
+    def max_frame_rate(self) -> float:
+        """Max frame rate in Hz for the current configuration"""
+        return 1 / self.minimum_frame_time
+
+    @max_frame_rate.setter
+    def max_frame_rate(self, mfr: float):
+        self.minimum_frame_time = 1 / mfr
+
+    @property
+    def minimum_frame_time(self) -> float:
+        sample_time = self.minimum_values_per_reading / self.base_sample_rate
+        return self.readings_per_frame * sample_time
+
+    @minimum_frame_time.setter
+    def minimum_frame_time(self, frame: float):
+        sample_time = self.minimum_values_per_reading / self.base_sample_rate
+        self.readings_per_frame = min(
+            self.maximum_readings_per_frame, int(floor(frame / sample_time))
+        )
+
+
+# TODO: need to change this name.
+MAX_CHANNELS = 11
+
+IDLE_TETRAMM = {
+    "drv.acquire": 0,
+}
+
+COMMON_TETRAMM = {
+    "drv.range": TetrammRange.uA,
+    "drv.num_channels": "4",
+    "drv.resolution": TetrammResolution.TwentyFourBits,
+    "drv.geometry": TetrammGeometry.Square,
+}
+
+TRIGGERED_TETRAMM = {
+    **COMMON_TETRAMM,
+    "drv.trigger_mode": TetrammTrigger.ExtTrigger,
+}
+
+FREE_TETRAMM = {
+    **COMMON_TETRAMM,
+    "drv.values_per_reading": 10,
+    "drv.averaging_time": 0.1,
+    "drv.trigger_mode": TetrammTrigger.FreeRun,
+}
+
+
+def triggered_tetramm(dev: TetrammDriver) -> Generator[Msg, None, None]:
+    sigs = walk_rw_signals(dev)
+    yield from set_signal_values(
+        sigs,
+        [
+            IDLE_TETRAMM,
+            TRIGGERED_TETRAMM,
+        ],
+    )
+
+
+def free_tetramm(dev: TetrammDriver) -> Generator[Msg, None, None]:
+    """Freerun the tetramm, setting it to acquire so diodes update."""
+    sigs = walk_rw_signals(dev)
+    yield from set_signal_values(sigs, [IDLE_TETRAMM, FREE_TETRAMM])
+    yield from bps.abs_set(dev.drv.acquire, 1)
+
+
+class TetrammShapeProvider(ShapeProvider):
+    def __init__(self, controller: TetrammController) -> None:
+        self.controller = controller
+
+    async def __call__(self) -> Sequence[int]:
+        return [MAX_CHANNELS, self.controller.readings_per_frame]
+
+
+class TetrammDetector(StandardDetector):
+    def __init__(
+        self,
+        prefix: str,
+        directory_provider: DirectoryProvider,
+        name: str,
+        **scalar_sigs: str,
+    ) -> None:
+        drv = TetrammDriver(prefix + "DRV:")
+        hdf = NDFileHDF(prefix + "HDF5:")
+
+        self.drv = drv
+        self.hdf = hdf
+
+        # TODO: how to make the below, readable signals?
+        # self.current_1 = ad_r(float, prefix + ":Cur1:MeanValue")
+        # self.current_2 = ad_r(float, prefix + ":Cur2:MeanValue")
+        # self.current_3 = ad_r(float, prefix + ":Cur3:MeanValue")
+        # self.current_4 = ad_r(float, prefix + ":Cur4:MeanValue")
+
+        # self.position_x = ad_r(float, prefix + ":PosX:MeanValue")
+        # self.position_y = ad_r(float, prefix + ":PosY:MeanValue")
```

Comment:
Can we either do the todo or create an issue for it and link the issue in the code comment?

Reply:
https://github.com/DiamondLightSource/dodal/issues/337

Made an issue and removed the commented out code for now.